lambda/translcore.ml

```diff
@@ -2201,6 +2205,89 @@ and transl_record_unboxed_product ~scopes loc env fields repres opt_init_expr =
       let exp = transl_exp ~scopes init_expr_sort init_expr in
       Llet(Strict, layout, init_id, init_id_duid, exp, lam)
 
+(* See [jane/doc/extensions/_02-unboxed-types/block-indices.md] *)
+and transl_idx ~scopes loc env ba uas =  (*  *)
+  let ua_to_pos (Uaccess_unboxed_field (_, lbl)) =
+    (* erase singleton unboxed products before lambda *)
+    if Array.length lbl.lbl_all == 1 then None else Some lbl.lbl_pos
+  in
+  let uas_path = List.filter_map ua_to_pos uas in
+  begin match ba with
+  | Baccess_block (_, idx) ->
+    let idx = transl_exp ~scopes Jkind.Sort.Const.for_idx idx in
+    begin match uas with
+    | [] -> idx
+    | Uaccess_unboxed_field (_, lbl) :: _ ->
+      (* Preserve the invariant that products have at least two elements *)
+      let base_sort =
+        if Int.equal (Array.length lbl.lbl_all) 1 then
+          lbl.lbl_sort
+        else
+          Jkind.Sort.Const.Product
+            (Array.to_list (Array.map (fun lbl -> lbl.lbl_sort) lbl.lbl_all))
+      in
+      let base_layout = layout env lbl.lbl_loc base_sort lbl.lbl_res in
+      let mbe = mixed_block_element_of_layout base_layout in
+      (* [uas_path] is a path into [mbe] *)
+      Lprim (Pidx_deepen (mbe, uas_path), [idx], (of_location ~scopes loc))
+    end
+  | Baccess_field (id, lbl) ->
+    check_record_field_sort id.loc lbl.lbl_sort;
+    begin match lbl.lbl_repres with
+    | Record_boxed _
+    | Record_float | Record_ufloat ->
+      (* Assert that all unboxed fields are of singleton records *)
+      List.iter
+        (fun (Uaccess_unboxed_field (_, l)) ->
+            if Array.length l.lbl_all <> 1 then
+              Misc.fatal_error "Texp_idx: non-singleton unboxed record field \
+                in non-mixed boxed record")
+        uas;
+      Lprim (Pidx_field lbl.lbl_pos, [], (of_location ~scopes loc))
+    | Record_inlined _ | Record_unboxed ->
+      Misc.fatal_error "Texp_idx: unexpected unboxed/inlined record"
+    | Record_mixed shape ->
+      let shape =
+        Lambda.transl_mixed_product_shape
+          ~get_value_kind:(fun _ -> Lambda.generic_value) shape
+      in
+      (* Check to make sure the gap never overflows.
+         See [jane/doc/extensions/_02-unboxed-types/block-indices.md]. *)
+      let cts =
+        Mixed_product_bytes_wrt_path.count_shape shape lbl.lbl_pos uas_path in
+      if Option.is_none
+          (Mixed_product_bytes_wrt_path.offset_and_gap cts) then
```

Comment:
Two tiny style nits, I think the usual thing is to put `in` on its own line if it does not fit on the same line as `let`, and to put `then` at the beginning of a line if it does not fit on the same line as `if`.